enhancements/storage/csi-migration.md

```diff
@@ -0,0 +1,200 @@
+
+It is important to notice that, with this carry-patch in OCP, Attach Detach Controller will _not_ change its current behaviour as long as nodes are not migrated to CSI, which is the default behaviour in OCP 4.8.
+
+#### Graduation
+
+During Tech Preview in OCP 4.8, users can enable the CSI migration using a *FeatureSet*. It could be shared with [CCMO](https://github.com/openshift/enhancements/pull/463), but we may want to have a specific *FeatureSet* only for CSI Migration for users that want to enable CSI Migration without migrating to an external cloud provider.
+
+Once CSI migration reaches GA in upstream, the associated feature gates will be enabled by default. As a result, it will not be necessary to use *FeatureSets* anymore.
+
+### Risks and Mitigations
+
+A carry-patch means that OCP will be the only Kubernetes distribution exercising this code path, which can lead us to bugs that were not seen anywhere. However, we are confident that the patch is small and self-contained enough to be used.
+
+## Design Details
+
+### Test Plan
+
+#### E2E jobs
+
+We want E2E jobs for all migrated plugins ready at **Tech Preview** time.
+
+For each E2E job:
+
+1. Install an OCP cluster.
+1. Enable the `CSIMigration` _FeatureSet_.
+1. Run E2E tests for in-tree volume plugins. This should use the CSI driver instead.
+1. Disable the `FeatureSet`.
+1. Once again, run E2E tests for in-tree volume plugins.
+
+In addition to that, as a stretch goal, we want a separate job that:
+
+1. Runs a `StatefulSet`.
+1. Enables the migration _FeatureSets_.
+1. Wait for all components to have the right feature flags.
+1. Checks if the StatefulSet survives.
+
+Once CSI migration is GA, we expect the regular upgrade jobs will cover upgrades from an OCP version with migration disabled to a version with migration enabled.
+
+### Graduation Criteria
+
+This is what needs to be done across different support phases:
+
+1. Tech Preview in OCP 4.8:
+
+* Introduce a new *FeatureSet* in openshift/api called `CSIMigration`.
+* Make sure the *FeatureSet* used by [CCMO](https://github.com/openshift/enhancements/pull/463) contains the CSI migration feature flags enabled for the respective storage backend.
+* Introduce an upstream patch that allows Attach Detach Controller to have its own custom feature gates, independent from Kube Controller Manager.
+* Introduce a carry-patch in OCP that enables CSI Migration for Cinder and GCP PD in Attach Detach Controller.
+* A PoC of both upstream and OCP patches [are available](https://github.com/openshift/kubernetes/pull/601).
+
+2. GA in OCP 4.9
+
+Nothing needs to be done, migration flags are enabled by default and cannot be disabled.
+
+3. Post-GA in OCP 4.10
+
+* Remove `CSIMigration` *FeatureSet* from openshift/api.
+
```

Comment:
Various in-tree plugins will migrate in various releases. We should note that here. The real migration will end in 4.11 - 4.12

Reply:
And we will need to update the carry patch to enable new CSI driver migration every release.

Reply:
Added